File: clients/apple/MultiLiveTV/Services/VodService.swift
import Foundation

@MainActor
final class VodService: ObservableObject {
    private let store: SourceStore

    init() {
        store = (try? SourceStore()) ?? SourceStore(sources: [])
        PlayLineWeighting.playPriorityBySourceId = Dictionary(
            uniqueKeysWithValues: store.all().map { ($0.numericId, $0.priority.playPriority) }
        )
    }

    func loadHomeLaunch(tvDisplay: Bool) async throws -> HomeLaunchPayload {
        try await Self.loadHomeLaunch(store: store, tvDisplay: tvDisplay)
    }

    nonisolated private static func loadHomeLaunch(
        store: SourceStore,
        tvDisplay: Bool
    ) async throws -> HomeLaunchPayload {
        try Task.checkCancellation()
        return try await HomeLaunch.load(
            fetchList: { slug in
                let result = try await CategoryListService.fetchUnifiedList(
                    store: store,
                    slug: slug,
                    page: 1
                )
                return HomeLaunch.ListPage(
                    items: VodMergeService.toVodItems(result.list),
                    page: result.page,
                    pageCount: result.pageCount
                )
            },
            tvDisplay: tvDisplay
        )
    }

    func fetchTypes(sourceId: Int? = nil) async throws -> TypesResponse {
        let source: Source
        if let sourceId {
            guard let found = store.byID(sourceId) else { throw VodError.sourceNotFound }
            source = found
        } else {
            guard let found = store.default() else { throw VodError.sourceNotFound }
            source = found
        }

        let types = try await SourceCollector.fetchTypes(source: source)
        let categories = MacCMSCategoryService.buildCategories(from: types)
        return TypesResponse(
            source: ListResponse.SourceRef(id: source.id, name: source.name),
            categories: categories
        )
    }

    func fetchList(
        page: Int,
        slug: String? = nil,
        hours: Int? = nil
    ) async throws -> ListResponse {
        let result = try await CategoryListService.fetchUnifiedList(
            store: store,
            slug: slug,
            page: page,
            hours: hours
        )
        return ListResponse(
            source: nil,
            typeId: nil,
            page: result.page,
            pagecount: result.pageCount,
            total: result.total,
            list: VodMergeService.toVodItems(result.list)
        )
    }

    /// 单源 type_id 列表（调试 / 兼容）
    func fetchListByType(
        page: Int,
        typeId: Int? = nil,
        sourceId: Int? = nil,
        knownChildTypeIds: [Int] = [],
        hours: Int? = nil
    ) async throws -> ListResponse {
        if let sourceId {
            guard let source = store.byID(sourceId) else { throw VodError.sourceNotFound }
            let result = try await CategoryListService.fetchVodListByTypeMerged(
                store: store,
                source: source,
                typeId: typeId,
                page: page,
                knownChildTypeIds: knownChildTypeIds,
                hours: hours
            )
            return makeListResponse(source: source, typeId: typeId, result: result)
        }

        for source in store.enabled() {
            do {
                let result = try await CategoryListService.fetchVodListByTypeMerged(
                    store: store,
                    source: source,
                    typeId: typeId,
                    page: page,
                    knownChildTypeIds: knownChildTypeIds,
                    hours: hours
                )
                if !result.list.isEmpty {
                    return makeListResponse(source: source, typeId: typeId, result: result)
                }
            } catch {
                continue
            }
        }

        guard let fallback = store.default() else { throw VodError.sourceNotFound }
        return ListResponse(
            source: ListResponse.SourceRef(id: fallback.id, name: fallback.name),
            typeId: typeId,
            page: page,
            pagecount: 0,
            total: 0,
            list: []
        )
    }

    func search(_ keyword: String, page: Int = 1) async throws -> [VodItem] {
        var mergeable: [MergeableVodItem] = []
        let searchable = store.collectable(capability: \.search)

        await withTaskGroup(of: [MergeableVodItem].self) { group in
            for source in searchable {
                group.addTask {
                    guard let pageResult = try? await SourceCollector.search(
                        source: source,
                        keyword: keyword,
                        page: page
                    ) else {
                        return []
                    }
                    return pageResult.list.map {
                        MergeableVodItem(
                            item: $0.toVodItemRaw(),
                            sourceId: source.id,
                            sourceName: source.name
                        )
                    }
                }
            }
            for await batch in group {
                mergeable.append(contentsOf: batch)
            }
        }

        let merged = VodMergeService.mergeVodItems(mergeable, store: store)
        return CategoryMatch.excludingHidden(VodMergeService.toVodItems(merged))
    }

    func detail(
        sourceId: Int,
        vodId: String,
        onPrimary: ((DetailResponse) -> Void)? = nil
    ) async throws -> DetailResponse {
        guard let source = store.byID(sourceId) else { throw VodError.sourceNotFound }
        guard let (primary, primaryResponse) = try await VodMergeService.fetchPrimaryVodDetail(
            source: source,
            vodId: vodId
        ) else {
            throw VodError.vodNotFound
        }
        onPrimary?(primaryResponse)
        return await VodMergeService.enrichMergedVodDetail(
            store: store,
            source: source,
            vodId: vodId,
            primary: primary
        )
    }

    func parsePlay(sourceId: Int, url: String) async throws -> ParseResponse {
        guard let source = store.byID(sourceId) else { throw VodError.sourceNotFound }
        let response = await PlayParser.parsePlayAddress(source: source, playURL: url)
        if PlaybackSupport.isDirectMediaURL(response.url) {
            return response
        }
        return ParseResponse(url: response.url, parsed: false)
    }

    func source(for id: Int) -> Source? {
        store.byID(id)
    }

    func fetchVodPic(sourceId: Int, vodId: String) async throws -> String? {
        guard let source = store.byID(sourceId) else { throw VodError.sourceNotFound }
        let movie = try await SourceCollector.detail(source: source, sourceMovieId: vodId)
        let pic = movie.poster
        return pic.isEmpty ? nil : pic
    }

    private func makeListResponse(source: Source, typeId: Int?, result: CategoryListService.ListResult) -> ListResponse {
        ListResponse(
            source: ListResponse.SourceRef(id: source.id, name: source.name),
            typeId: typeId,
            page: result.page,
            pagecount: result.pageCount,
            total: result.total,
            list: VodMergeService.toVodItems(result.list)
        )
    }
}

extension VodService: PlayURLParsing {}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/vod/VodService.kt
package com.heibaimiao.multilivetv.vod

import com.heibaimiao.multilivetv.category.CategoryListService
import com.heibaimiao.multilivetv.category.CategoryMatch
import com.heibaimiao.multilivetv.category.MacCMSCategoryService
import com.heibaimiao.multilivetv.home.HomeLaunch
import com.heibaimiao.multilivetv.home.HomeLaunchPayload
import com.heibaimiao.multilivetv.maccms.MergeableVodItem
import com.heibaimiao.multilivetv.merge.VodMergeService
import com.heibaimiao.multilivetv.model.DetailResponse
import com.heibaimiao.multilivetv.model.ListResponse
import com.heibaimiao.multilivetv.model.ParseResponse
import com.heibaimiao.multilivetv.model.Source
import com.heibaimiao.multilivetv.model.TypesResponse
import com.heibaimiao.multilivetv.model.VodError
import com.heibaimiao.multilivetv.model.VodItem
import com.heibaimiao.multilivetv.model.VodVariant
import com.heibaimiao.multilivetv.net.VodClientException
import com.heibaimiao.multilivetv.parser.PlayLineWeighting
import com.heibaimiao.multilivetv.parser.PlayParser
import com.heibaimiao.multilivetv.parser.PlaybackSupport
import com.heibaimiao.multilivetv.source.SourceCapability
import com.heibaimiao.multilivetv.source.SourceCollector
import com.heibaimiao.multilivetv.source.SourceStore
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

class VodService(private val store: SourceStore = runCatching { SourceStore() }.getOrElse { SourceStore(emptyList()) }) {
    init {
        PlayLineWeighting.playPriorityBySourceId =
            store.all().associate { it.numericId to it.priority.playPriority }
    }

    suspend fun loadHomeLaunch(tvDisplay: Boolean = false): HomeLaunchPayload =
        HomeLaunch.load(
            fetchList = { slug ->
                val result = CategoryListService.fetchUnifiedList(store, slug, page = 1)
                HomeLaunch.ListPage(VodMergeService.toVodItems(result.list), result.page, result.pageCount)
            },
            tvDisplay = tvDisplay,
        )

    suspend fun fetchTypes(sourceId: Int? = null): TypesResponse {
        val source = sourceId?.let { store.byID(it) } ?: store.default()
            ?: throw VodClientException(VodError.SOURCE_NOT_FOUND.message)
        val types = SourceCollector.fetchTypes(source)
        return TypesResponse(
            source = ListResponse.SourceRef(source.id, source.name),
            categories = MacCMSCategoryService.buildCategories(types),
        )
    }

    suspend fun fetchList(page: Int, slug: String? = null, hours: Int? = null): ListResponse {
        val result = CategoryListService.fetchUnifiedList(store, slug, page, hours)
        return ListResponse(
            source = null,
            typeId = null,
            page = result.page,
            pagecount = result.pageCount,
            total = result.total,
            list = VodMergeService.toVodItems(result.list),
        )
    }

    suspend fun search(keyword: String, page: Int = 1): List<VodItem> = coroutineScope {
        val searchable = store.collectable(SourceCapability.SEARCH)
        val mergeable = searchable.map { source ->
            async {
                val pageResult = runCatching { SourceCollector.search(source, keyword, page) }.getOrNull()
                pageResult?.list.orEmpty().map {
                    MergeableVodItem(it.toVodItemRaw(), source.id, source.name)
                }
            }
        }.awaitAll().flatten()
        val merged = VodMergeService.sortMergedByUpdatedDesc(
            VodMergeService.mergeVodItems(mergeable, store),
        )
        CategoryMatch.excludingHidden(VodMergeService.toVodItems(merged))
    }

    suspend fun detail(
        sourceId: Int,
        vodId: String,
        knownVariants: List<VodVariant> = emptyList(),
        onPrimary: ((DetailResponse) -> Unit)? = null,
    ): DetailResponse {
        val source = store.byID(sourceId) ?: throw VodClientException(VodError.SOURCE_NOT_FOUND.message)
        val fetched = VodMergeService.fetchPrimaryVodDetail(source, vodId)
            ?: throw VodClientException(VodError.VOD_NOT_FOUND.message)
        onPrimary?.invoke(fetched.second)
        return VodMergeService.enrichMergedVodDetail(
            store = store,
            source = source,
            vodId = vodId,
            primary = fetched.first,
            knownVariants = knownVariants,
        )
    }

    suspend fun parsePlay(sourceId: Int, url: String): ParseResponse {
        val source = store.byID(sourceId) ?: throw VodClientException(VodError.SOURCE_NOT_FOUND.message)
        val response = PlayParser.parsePlayAddress(source, url)
        return if (PlaybackSupport.isDirectMediaURL(response.url)) response else ParseResponse(response.url, parsed = false)
    }

    fun source(id: Int): Source? = store.byID(id)
}
